### ISSUE TITLE
Method selection with null arguments is not right....

### ISSUE THREAD

**Issue by sourceforge-issue-exporter** [beanshell/beanshell#132]
Method selection doesn't catch when selections with null arguments are ambiguous... instead it pushes selection to the most derived match.

Reported by: patn

Original Ticket: "beanshell/bugs/11":https://sourceforge.net/p/beanshell/bugs/11

**Comment by nickl-** [comment]
**Closed: Resolved as per unit tests #552**

**Comment by opeongo** [comment]
In Java 16 String.valueOf(null) will call new String(char[]).

**Comment by nickl-** [comment]
> Unfortunately BeanShell only uses the type of the object being passed and does not look at the declared type of the variable.

Correct since there is no guarantee that it was declared of type.

This seems difficult, not sure we even have the variable declaration type at the point of method resolution.

This issue was regarding NullPointerException being thrown, and if I understand your analysis correctly we now return null without exception, is this correct? If so, does it really matter which method handled the null? ...as long as the non null values are handled correctly.

If we are not causing NullPointerException anymore then this issue can be closed, yes? Oh I see it is closed...hmmm.

**Comment by opeongo** [comment]
The problem is that at the moment a null argument cannot carry any type, so when beanshell does the method selection it is prone to having an ambiguous method selection problem that it resolves differently than Java.  

The `String.valueOf()` example is an easy on to look at because there are only two method options:
```java
String.valueOf(Object obj)
String.valueOf(char[] value)
```

If there is just a plain unadorned null with no type then according to the method selection rules `char[]` is more specific than `Object`, so the `String.valueOf(char[] value)` method should be selected.  This will throw an NPE when passed a null.  This is what Java does.

However, the deck is stacked [by these lines](https://github.com/beanshell/beanshell/blob/cb552a4d08c279680d6463b61e2f034fe12a590b/src/main/java/bsh/Types.java#L310-L313) in BeanShell that say if the argument is null then give it the type of String.class.  This is a weird trick to always make nulls match to String arguments.  As you can expect it works occasionally but fails just as often.  Because of this `String.valueOf(null)` always returns `"null"`.

If this was Java (statically typed) then the compiler most often than not would know the type and be able to make the correct method selection:
```java
Object a = null;
assert(String.valueOf(a).equals("null"));
assert(String.valueOf((Object)null).equals("null"));
```

So the fundamental problem is that BeanShell does not carry the type of the parameter when it is null.  BeanShell should be able to successfully handle this following case but it doesn't, it returns the string `"null"` because the current default type for null is String.class regardless of information to the contrary:
```java
try {
  String.valueOf((char[]) null);
  assert(false);  // Gets to here but it shouldn't
} catch (NullPointerException npe) {
    // Should come here, but currently it doesn't
}
```

You might say 'big deal' but this does cause problems.  Because of this method selection is different than Java, not a good thing.  Nulls are part of the language and sometimes API's allow null values.  If can be a big problem when the wrong method is called because of incorrect method selection, as in the following API, what method will be called when a null is passed in?

```java
public class MAD {
  public void invoke(Double delay) {
    System.out.println("Launch the missiles in "+delay+" seconds");
    // etc...
  }
  public void invoke(String message) {
    System.out.println("Peace on Earth, disarming due to "+message);
    // etc...
  }
}
```
Obviously a contrived example and perhaps not persuasive, but I have been bitten by the inaccurate selectivity.

I think that the proper way to deal with this problem is twofold:
1. Carry the inferred type for values that are null.  Instead of null being a [static Primitive.NULL](https://github.com/beanshell/beanshell/blob/cb552a4d08c279680d6463b61e2f034fe12a590b/src/main/java/bsh/Primitive.java#L96-L100) it should be a sub-class of Primitive that carries it's declared type (if available).  Something like:
```java
public class Null extends Primitive {
  public Null(Class clazz) {
    this.value = clazz;
  }
}
```
then add something like these methods to [Primitive.java](https://github.com/beanshell/beanshell/blob/cb552a4d08c279680d6463b61e2f034fe12a590b/src/main/java/bsh/Primitive.java)
```java
public isNull() {
  return this instanceof Null;
}
public Class<?> getType()
    {
        if ( this == Primitive.VOID )
            return Void.TYPE;

        // NULL return saved type
        if (this instanceof Null)
            return (Class) value;

        return unboxType( value.getClass() );
    }
```
The handling of `Primitive.NULL` will have to change throughout the API match this change, but this will allow a lightweight means to carry along type declarations and casts on null values.  There will be cases where the class isn't known but I would expect these to be uncommon wrt ambiguous methods and an exception would be thrown.  When they occur in BeanShell scripts exceptions can be avoided by explicit declarations or casts (the right way to go). 

2. Just like Java does, detect and throw an exception when faced with an ambiguous method selection.  At the moment there is no guarantee for what method is returned when more than one method matches (first encountered? idk).  But BeanShell is behaving differently than Java in this regard and it is not good.  

So tightening up how type information is carried with nulls, throwing an exception when faced with an ambiguous method selection, and capturing the type of casts on nulls in order to resolve ambiguities seems to be the proper way to go.

I am not sure it anything needs to be done about `Primitive.VOID`, I will look in to this also.

Let me know what you think and I will investigate further and prepare a prototype

**Comment by nickl-** [comment]
> The problem is that at the moment a null argument cannot carry any type, so when beanshell does the method selection it is prone to having an ambiguous method selection problem that it resolves differently than Java.

tldr; but I agree with the first line so far.

We should create a new issue, instead of having long discussions here in a closed thread arguably off topic, and in so doing add it to the list of things to do. 

Interesting problem, will return to this and read your analysis with deserved attention. Please reference this issue when creating the new one so it is easy to find.
